File: Back/.Net/Modules/Users/BetFriends.Users.Infrastructure/Repositories/Sql/SqlUserRepository.cs
using BetFriends.Shared.Infrastructure.Event;
using BetFriends.Users.Domain.Users;
using BetFriends.Users.Infrastructure.Repositories.Sql.DataAccess;
using Microsoft.EntityFrameworkCore;

namespace BetFriends.Users.Infrastructure.Repositories.Sql;

public class SqlUserRepository(UserContext userContext,
                                 DomainEventsAccessor domainEventsAccessor) : IUserRepository
{
    private readonly UserContext userContext = userContext;
    private readonly DomainEventsAccessor domainEventsAccessor = domainEventsAccessor;

    public Task<UserEntity?> GetEntityById(Guid id)
    {
        return userContext.Users.FirstOrDefaultAsync(x => x.Id == id);
    }

    public async Task<bool> IsUserExistAsync(string username, string email)
    {
        var entity = await userContext.Users.FirstOrDefaultAsync(x => x.Username == username ||
                                                                    x.Email == email);
        return entity != null;
    }

    public async Task SaveAsync(User user)
    {
        var entity = new UserEntity(user.Snapshot);
        await userContext.Users.AddAsync(entity);
        domainEventsAccessor.Add(user.Events);
    }

    public void SaveEntity(UserEntity entity)
    {
        userContext.Users.Add(entity);
        userContext.SaveChanges();
    }
}

File: Back/NodeJS/modules/users/src/infrastructure/repositories/SqlUserRepository.ts
import { DomainEventAccessor } from "../../../../shared/infrastructure/events/DomainEventAccessor";
import { IUserRepository } from "../../domain/users/IUserRepository";
import { User } from "../../domain/users/User";
import { UserContext } from "../dataAccess/UserContext";
import { UserEntity } from "../dataAccess/UserEntity";

export class SqlUserRepository implements IUserRepository {
    
    constructor(private readonly userContext: UserContext,
                private readonly domainEventAccessor: DomainEventAccessor
    ) {}
    async IsExists(email: string, userId: string, username: string): Promise<boolean> {
        const users = await this.userContext.UserEntity.find({
            where: [
                { Email: email },
                { Id: userId },
                { Username: username },
            ]
        })
        return users.length > 0;
    }
    async Save(user: User): Promise<void> {
        const entity = UserEntity.Create(user.GetSnapshot());
        await this.userContext.UserEntity.save(entity);
        this.domainEventAccessor.AddEvents(user.DomainEvents)
    }

    async GetEntityById(userId: string): Promise<UserEntity> {
        return await this.userContext.UserEntity.findOneBy({
            Id: userId
        })
    }

}